### PR TITLE
Move TileMapLayer to its own files

### PR DESCRIPTION
**Pull request by groud**
As  I mentioned in [this proposal](https://github.com/godotengine/godot-proposals/issues/6415), I'd like to move TileMap layers to their own node. This fits the 

I prefer to go there step by step. This first PR moves TileMap layer to another file. 

The next steps would be:
- Change TileMap layers to be used as an internal nodes for the TileMap node (and make sure everything works).
- Implement a TileMapGroup node, that would constraint its child TileMapLayer nodes to use a given TileSet, and to keep their transform. It should have a very simple API. Keep TileMapGroup node private for now.
- Make TileMap expand the TileMapGroup node API. 
- To ensure compatibility, implement a tool to transform the TileMap node into the (TileMapGroup + Children TileMap layers).
- At the same time: 
	- expose the TileMapLayer node.
	- expose the TileMapGroup node.
	- Change the editor to edit TileMapLayer directly (and, for simplicity of the implementation, probably disables it for the TileMap node. That kind of forces people to move the to new workflow by using the auto-conversion, but it's probably acceptable IMO)

Basically, the TileMap node could become the equivalent of (TileMapGroup + Children TileMap layers), with layers being internal nodes instead of being explicit ones.

### CODE REVIEW

File: scene/2d/tile_map.h [hunk 1]
```diff
@@ -31,409 +31,14 @@
 #ifndef TILE_MAP_H
 #define TILE_MAP_H
 
+#include "scene/2d/tile_map_layer.h"
+class TileMapLayer;
+class TerrainConstraint;
```
Comment (AThousandShips) [inline scene/2d/tile_map.h:36]: Aren't these classes included just before?

**Comment by AThousandShips** [comment]
I'm a bit concerned by the circular inclusion, can this not be handled in any other way?

**Comment by groud** [comment]
> I'm a bit concerned by the circular inclusion, can this not be handled in any other way?

I thought there was not, but I managed to solve it.

File: scene/2d/tile_map_layer.h [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef TILE_MAP_LAYER_H
+#define TILE_MAP_LAYER_H
+
+#include "scene/2d/node_2d.h"
+#include "scene/2d/tile_map.h"
+#include "scene/gui/control.h"
```
Comment (KoBeWi) [inline scene/2d/tile_map_layer.h:36]: What is `node_2d` and `control` for?

File: scene/2d/tile_map.h [hunk 3]
```diff
@@ -35,403 +35,14 @@
 #include "scene/gui/control.h"
```
Comment (KoBeWi) [inline scene/2d/tile_map.h:35]: Unrelated, but this can be forward-declared.

Comment (akien-mga) [reply scene/2d/tile_map.h:35]: Forward declaring in headers helps reduce the amount of files that need to be rebuilt on incremental builds. If a commit changes `control.h`, `tile_map.h` and all the files that include it would need to be recompiled.

If it's forward declared, only `tile_map.cpp` needs to be recompiled.

Comment (groud) [reply scene/2d/tile_map.h:35]: Oh ok, that makes sense! I'll update it then, thanks.

**Review by KoBeWi — APPROVED** [review]

**Comment by groud** [comment]
> Needs a rebase and we can merge it soon if it unblocks your further work.

I'll rebase once https://github.com/godotengine/godot/pull/84660 is merged, to avoid conflicts.

**Comment by YuriSizov** [comment]
@groud You should be able to rebase now!

**Comment by akien-mga** [comment]
New Year's rebase poke ;)

**Comment by groud** [comment]
> New Year's rebase poke ;)

Done!

**Comment by akien-mga** [comment]
Thanks!
